tonconnect-server/TonConnect2.md

```diff
@@ -0,0 +1,533 @@
+
+Decentralized apps authenticate users within smart contracts using message sender addresses. 
+The UI (“offchain logic”) runs on behalf of the user, so there is no need for a dedicated “sign in” scheme.
+
+Centralized apps run on the servers and may store user data. They may use “sign in” functionality to authenticate the user, where instead of username/password pair we use cryptographic keypair derived from the wallet secret seed.
+
+
+### App authentication
+
+Apps are authenticated using [TON DNS](https://ton.org/docs/#/web3/dns) protocol.
+
+Decentralized apps may use `sha256("wallet")` key for the smart contract address. 
+When the wallet receives request to sign a message to a smart contract with a given **.ton** name, a valid and up-to-date DNS record would be used to authenticate that contract address with that name.
+
+**TODO:** callback/return URLs should also be specified in the TON DNS to permit their verification for dapps that cannot sign their own requests.
+
+Centralized services may also use `sha256("tonconnect.app-pk")` key to specify 32-byte [app public key](#app-keypair) used to initiate authentication requests to the wallet. The key should be stored on the online server and can be frequently rotated via TON.DNS record update.
+
+
+
+## Workflows
+
+### Connect workflow
+
+1. [App](#app) or [service](#service) prepares [ConnectRequest](#ConnectRequest) and displays a QR code and a button to let the [wallet](#wallet) connect.
+2. User scans the QR or clicks the button.
+3. Wallet shows confirmation dialog that shows the wallet address communicated to the app.
+4. Wallet stores app’s web address (if provided) in the list of connected services.
+5. Wallet sends/redirects the user back to the app with the reply: [ConnectResponse](#ConnectResponse) containing its address and optional push URL.
+6. App initializes its UI using the provided address.
+
+### Service authorization
+
+This is for “offchain” operations: where the wallet responds with authorization to a service directly instead of signing a transaction.
+
+1. User presses a button in the [service](#service).
+2. Service requests authorization of some action by the user.
+3. Wallet shows confirmation dialog with the details of the authorization.
+4. User confirms.
+5. Wallet sends/redirects the user back to the app with the reply.
+6. Service performs the requested action (updates its database, makes blockchain transactions of its own etc.).
+
+
+### Transaction authorization
+
+This is for the “onchain” operations: transactions signed by the wallet and sent by the user directly from their wallet.
+
+1. User presses a button in the [app](#app).
+2. App/service requests authorization of some action by the user.
+3. Wallet shows confirmation dialog with the details of the authorization.
+4. User confirms.
+5. Wallet publishes the transaction.
+6. Wallet sends/redirects the user back to the app/service with the reply.
+7. App/service publishes the transaction.
+
+
```

Comment:
We're missing a workflow for an offchain signature without any *service* in existence.

This workflow exists on Ethereum and has proven quite useful: https://web3js.readthedocs.io/en/v1.2.11/web3-eth.html?#sign

For example, when my bank did due diligence over my wallet addresses, it requested that I sign a message with my details to prove ownership of these wallet addresses. There are also DAOs that allow you to cast votes through a similar mechanism, you vote by signing a JSON of your vote.

These use cases can run fully client side and do not require a centralized backend to be involved. We should support them.

Reply:
Another flow that must be supported under this use-case is to be able to easily verify this signature from a FunC contract. So the full flow will be: End-user uses their mobile app wallet to sign a short cell off-chain, the client code gets a signature back. This signature is sent later (possibly by a different party) to a FunC smart contract. The contract can verify the signature and be convinced that the address (public key?) of the user is indeed the one who signed.

Reply:
Yeah, we probably should adopt BoC-safe prefix pioneered by Tonhub: https://github.com/ton-foundation/specs/blob/main/specs/wtf-0002.md

A few thoughts:
1. Ideally we have one format for custom signature, and if we want to be TVM-compatible, that should be through BoC/TL-B instead of TL.
2. We need secure domain separation, possibly tied to TON.DNS, or to a pubkey of the recipient with encryption and stuff. For fully offchain use we already have "AuthRequest" flow that's tied to a proof of ownership of the address. It is interactive/ephemeral, though.

Reply:
Yeah, we probably should adopt BoC-safe prefix pioneered by Tonhub: https://github.com/ton-foundation/specs/blob/main/specs/wtf-0002.md

A few thoughts:
1. Ideally we have one format for custom signature, and if we want to be TVM-compatible, that should be through BoC/TL-B instead of TL.
2. We need secure domain separation, possibly tied to TON.DNS, or to a pubkey of the recipient with encryption and stuff. For fully offchain use we already have "AuthRequest" flow that's tied to a proof of ownership of the address. It is interactive/ephemeral, though.

Reply:
@talkol isn't on-chain authentication (for voting/administration) through internal msg superior to signature checks? This way you move out costs of sig checks to the individual wallets and have uniform "sender address" as authentication mechanism. Then you enable implementation of some policies through those addresses. E.g. one of the voters might be a multisig contract, or a thing with some ACLs or timelocks.

Reply:
@oleganza on-chain authentication is impossible for certain use-cases and adding the off-chain flexibility is important. One big reason is gas. Some use-cases require gas-less operations since users are not willing to pay for them, even a few cents. One popular example is snapshot.org, getting DAO participants to vote is so difficult, that if each one of them had to pay something to vote, participation would drop by 50%.

I would be very careful designing a standard that limits developers from doing certain things. I don't think we can predict what developers of future apps will require or not. Off-chain signatures are a basic requirement that is available everywhere.

Reply:
> For fully offchain use we already have "AuthRequest" 

From what I understand, AuthRequest requires a service to be involved, meaning an entity that holds a private key. The use-cases I want to support do not permit any service to be part of the flow.

A good example is signing something and sending the bank as proof for them to check manually. Example on [Ethereum](https://support.mycrypto.com/how-to/getting-started/how-to-sign-and-verify-messages-on-ethereum/). Why would I want a third-party service provider to be involved with something like this?